### PR TITLE
Migrate acos to structured kernel

### LINKED ISSUE #55070: Port kernels to be structured [tracker]
**WARNING: BEFORE STARTING ON A STRUCTURED KERNEL PORT, CHECK OPEN PULL REQUESTS TO SEE IF SOMEONE IS WORKING ON IT ALREADY.** https://github.com/pytorch/pytorch/pulls?q=is%3Apr+is%3Aopen+structured+in%3Atitle

RFC at: https://github.com/pytorch/rfcs/blob/rfc-0005/RFC-0005-structured-kernel-definitions.md

Here is the list of all functions which are immediately eligible to be ported to structured (as in, they have both a functional and out version with exactly the same signatures):

- [ ] abs (Anjali)
- [ ] absolute
- [ ] angle
- [ ] sgn
- [ ] conj
- [x] acos https://github.com/pytorch/pytorch/pull/56082
- [ ] arccos
- [x] add.Tensor
- [ ] _add_relu.Tensor
- [x] addmv (#55746)
- [ ] addr
- [ ] all.dim
- [ ] all.dimname
- [ ] any.dim
- [ ] any.dimname
- [ ] argmax
- [ ] argmin
- [x] acosh (#55540)
- [ ] arccosh
- [x] asinh https://github.com/pytorch/pytorch/pull/56082
- [ ] arcsinh
- [x] atanh
- [ ] arctanh
- [x] asin
- [ ] arcsin
- [x] atan https://github.com/pytorch/pytorch/pull/56082
- [ ] arctan
- [ ] baddbmm https://github.com/pytorch/pytorch/pull/64805
- [ ] bernoulli
- [ ] binary_cross_entropy
- [ ] binary_cross_entropy_backward
- [ ] bitwise_not
- [ ] copysign.Tensor #55040
- [ ] logical_not
- [ ] logical_xor
- [ ] logical_and
- [ ] logical_or
- [ ] bmm https://github.com/pytorch/pytorch/pull/64805
- [ ] _bmm
- [ ] cat
- [ ] cat.names
- [x] ceil (#57589)
- [x] clamp (#61361)
- [ ] clamp_max
- [ ] clamp_min
- [ ] clip
- [ ] complex
- [ ] polar
- [x] cos (#55564)
- [x] cosh (#55563)
- [ ] cummax
- [ ] cummax.dimname
- [ ] cummin
- [ ] cummin.dimname
- [ ] cumprod
- [ ] cumprod.dimname
- [ ] cumsum
- [ ] cumsum.dimname
- [ ] diff
- [ ] div.Tensor
- [ ] div.Tensor_mode
- [ ] divide.Tensor
- [ ] divide.Tensor_mode
- [ ] true_divide.Tensor
- [ ] dot
- [ ] vdot
- [ ] row_stack
- [x] erf https://github.com/pytorch/pytorch/pull/56082
- [x] erfc https://github.com/pytorch/pytorch/pull/56082
- [x] exp https://github.com/pytorch/pytorch/pull/56082
- [x] exp2 https://github.com/pytorch/pytorch/pull/56082
- [x] expm1 https://github.com/pytorch/pytorch/pull/56082
- [x] floor (#57587)
- [ ] floor_divide
- [ ] frac
- [x] gcd (#57624)
- [x] lcm (#57628)
- [ ] _fft_r2c
- [ ] _fft_c2r
- [ ] _fft_c2c (#55730)
- [ ] inverse
- [ ] kron
- [ ] kthvalue
- [ ] kthvalue.dimname
- [ ] nan_to_num
- [ ] ldexp.Tensor
- [x] log https://github.com/pytorch/pytorch/pull/56082
- [x] log10 https://github.com/pytorch/pytorch/pull/56082
- [x] log1p https://github.com/pytorch/pytorch/pull/56082
- [x] log2 https://github.com/pytorch/pytorch/pull/56082
- [x] logaddexp (#57629)
- [x] logaddexp2 (#57629)
- [ ] xlogy.Tensor (#60814)
- [ ] xlogy.Scalar_Self (#60814)
- [ ] xlogy.Scalar_Other (#60814)
- [ ] _logcumsumexp
- [ ] logcumsumexp
- [ ] logcumsumexp.dimname
- [ ] _log_softmax (#57374)
- [ ] _log_softmax_backward_data @SplitInfinity 
- [ ] logsumexp
- [ ] logsumexp.names
- [ ] matmul
- [ ] matrix_power
- [ ] _compute_linear_combination
- [ ] max.dim
- [ ] max.names_dim
- [ ] amax
- [ ] mean.dim
- [ ] mean.names_dim
- [ ] median.dim
- [ ] median.names_dim
- [ ] nanmedian.dim
- [ ] nanmedian.names_dim
- [ ] min.dim
- [ ] min.names_dim
- [ ] amin
- [x] mm (#57755)
- [ ] mode
- [ ] mode.dimname
- [ ] mul.Tensor
- [ ] multiply.Tensor
- [ ] mv
- [ ] narrow_copy
- [ ] native_batch_norm
- [ ] batch_norm_elemt
- [ ] rad2deg
- [ ] deg2rad
- [x] reciprocal https://github.com/pytorch/pytorch/pull/56082
- [x] neg https://github.com/pytorch/pytorch/pull/57212
- [ ] negative
- [x] round
- [ ] rsqrt
- [x] silu (#58050)
- [ ] sigmoid https://github.com/pytorch/pytorch/pull/56082
- [ ] logit
- [x] sin
- [x] sinc
- [x] sinh (#55538)
- [ ] sspaddmm
- [ ] stack
- [ ] _stack
- [ ] hstack
- [ ] vstack
- [ ] dstack
- [ ] sum.dim_IntList
- [ ] sum.dim_DimnameList
- [ ] nansum.dim_IntList
- [ ] sqrt https://github.com/pytorch/pytorch/pull/56082
- [ ] square (#58266)
- [ ] std.dim
- [ ] std.names_dim
- [ ] prod.dim_int
- [ ] prod.dim_Dimname
- [x] tan https://github.com/pytorch/pytorch/pull/56082
- [x] tanh https://github.com/pytorch/pytorch/pull/56082
- [ ] tensordot
- [ ] threshold
- [x] trunc (#57350)
- [ ] fix
- [ ] var.dim
- [ ] var.names_dim
- [ ] norm.ScalarOpt_dim_dtype
- [ ] norm.ScalarOpt_dim
- [ ] norm.names_ScalarOpt_dim_dtype
- [ ] norm.names_ScalarOpt_dim
- [ ] frexp.Tensor
- [ ] frobenius_norm.dim
- [ ] nuclear_norm
- [ ] nuclear_norm.dim
- [ ] sub.Tensor
- [ ] subtract.Tensor
- [ ] heaviside
- [x] addmm (#57417)
- [ ] hspmm
- [ ] eq.Scalar (#60177)
- [ ] eq.Tensor (#60177)
- [ ] bitwise_and.Tensor (#60813)
- [ ] bitwise_and.Scalar (#60813)
- [ ] bitwise_or.Tensor 
- [ ] bitwise_or.Scalar
- [ ] bitwise_xor.Tensor (#60812)
- [ ] bitwise_xor.Scalar (#60812)
- [x] atan2 #55130
- [x] tril
- [x] triu
- [x] digamma https://github.com/pytorch/pytorch/pull/56082
- [ ] renorm
- [x] lerp.Scalar
- [x] lerp.Tensor
- [ ] fmod.Scalar (#60809)
- [ ] fmod.Tensor (#60809)
- [ ] remainder.Scalar
- [ ] remainder.Tensor
- [ ] addbmm (#60647)
- [ ] addcdiv
- [ ] diag
- [ ] cross
- [ ] ne.Scalar
- [ ] ne.Tensor
- [ ] not_equal.Scalar
- [ ] not_equal.Tensor
- [ ] ge.Scalar
- [ ] ge.Tensor
- [ ] greater_equal.Scalar
- [ ] greater_equal.Tensor
- [ ] le.Scalar
- [ ] le.Tensor
- [ ] less_equal.Scalar
- [ ] less_equal.Tensor
- [ ] gt.Scalar
- [ ] gt.Tensor
- [ ] greater.Scalar
- [ ] greater.Tensor
- [ ] lt.Scalar
- [ ] lt.Tensor
- [ ] less.Scalar
- [ ] less.Tensor
- [ ] take
- [ ] take_along_dim
- [ ] index_select
- [ ] index_select.dimname
- [ ] masked_select
- [ ] nonzero (@krshrimali)
- [ ] gather
- [ ] gather.dimname
- [ ] addcmul
- [ ] lstsq
- [ ] triangular_solve https://github.com/pytorch/pytorch/pull/61857
- [ ] symeig
- [ ] eig
- [ ] svd
- [ ] cholesky
- [ ] cholesky_solve
- [ ] solve
- [ ] cholesky_inverse
- [ ] qr
- [ ] geqrf
- [ ] orgqr
- [ ] ormqr
- [ ] lu_solve
- [ ] multinomial
- [x] lgamma https://github.com/pytorch/pytorch/pull/56082
- [ ] polygamma
- [x] erfinv https://github.com/pytorch/pytorch/pull/56082
- [ ] i0
- [x] sign (#57588)
- [ ] signbit (#57936)
- [ ] histc
- [x] hypot (#57627)
- [x] igamma (#57626)
- [x] igammac (#57626)
- [x] nextafter (#57625)
- [x] fmin
- [x] fmax
- [x] maximum (#57630)
- [ ] max.other
- [x] minimum (#57630)
- [ ] min.other
- [ ] quantile.scalar
- [ ] quantile
- [ ] nanquantile.scalar
- [ ] nanquantile
- [ ] sort
- [ ] sort.stable
- [ ] sort.dimname
- [ ] sort.dimname_stable
- [ ] msort
- [ ] topk
- [ ] pow.Tensor_Tensor
- [ ] pow.Scalar
- [ ] pow.Tensor_Scalar
- [ ] float_power.Tensor_Tensor - [#60855]
- [ ] float_power.Scalar - [#60855]
- [ ] float_power.Tensor_Scalar = [#60855]
- [ ] normal.Tensor_float
- [ ] normal.float_Tensor
- [ ] normal.Tensor_Tensor
- [ ] _cumsum
- [ ] _cumprod
- [ ] _cat
- [ ] _mode
- [ ] bucketize.Tensor
- [ ] searchsorted.Tensor
- [ ] mse_loss
- [ ] mse_loss_backward
- [ ] l1_loss
- [ ] l1_loss_backward
- [ ] multi_margin_loss
- [ ] multi_margin_loss_backward
- [ ] multilabel_margin_loss
- [ ] multilabel_margin_loss_forward
- [ ] multilabel_margin_loss_backward
- [ ] nll_loss
- [ ] nll_loss_forward
- [ ] nll_loss_backward
- [ ] nll_loss2d
- [ ] nll_loss2d_forward
- [ ] nll_loss2d_backward
- [x] smooth_l1_loss (#67404)
- [ ] smooth_l1_loss_backward
- [ ] huber_loss
- [ ] huber_loss_backward
- [ ] soft_margin_loss
- [ ] soft_margin_loss_backward
- [x] elu (#57619)
- [ ] glu
- [ ] glu_backward
- [x] hardsigmoid (#57622)
- [ ] hardtanh
- [ ] hardtanh_backward
- [ ] hardswish (this one is complicated, see https://github.com/pytorch/pytorch/pull/66899 )
- [x] leaky_relu (#57621)
- [ ] log_sigmoid
- [ ] log_sigmoid_forward
- [ ] log_sigmoid_backward
- [ ] rrelu_with_noise
- [x] softplus (#57620)
- [x] softplus_backward (#58482)
- [x] softshrink (#57623)
- [ ] softshrink_backward
- [ ] (XLA) adaptive_avg_pool2d (this one is a little tricky because there is specialized derivative handling; see ngimel)
- [ ] adaptive_avg_pool3d
- [ ] adaptive_avg_pool3d_backward
- [x] adaptive_max_pool2d (#56317)
- [x] adaptive_max_pool2d_backward (#56799)
- [x] adaptive_max_pool3d (#56320)
- [x] adaptive_max_pool3d_backward (#56800)
- [x] avg_pool2d (#58987)
- [ ] avg_pool2d_backward
- [ ] avg_pool3d
- [ ] avg_pool3d_backward (#59084)
- [x] fractional_max_pool2d
- [ ] fractional_max_pool2d_backward
- [ ] fractional_max_pool3d
- [ ] fractional_max_pool3d_backward
- [x] max_pool2d_with_indices
- [x] max_pool2d_with_indices_backward
- [ ] max_pool3d_with_indices
- [ ] max_pool3d_with_indices_backward
- [ ] max_unpool2d
- [ ] max_unpool2d_backward
- [ ] max_unpool3d
- [ ] max_unpool3d_backward
- [x] reflection_pad1d - #55531
- [x] reflection_pad1d_backward - #59103
- [ ] reflection_pad2d
- [ ] reflection_pad2d_backward
- [x] replication_pad1d - #55481
- [x] replication_pad1d_backward - ezyang
- [x] replication_pad2d - asuhan
- [ ] replication_pad2d_backward
- [x] replication_pad3d - ezyang
- [ ] replication_pad3d_backward
- [x] upsample_linear1d
- [x] upsample_linear1d_backward
- [x] upsample_bilinear2d
- [x] upsample_bilinear2d_backward
- [x] upsample_bicubic2d
- [x] upsample_bicubic2d_backward
- [x] upsample_trilinear3d
- [x] upsample_trilinear3d_backward
- [x] upsample_nearest1d
- [x] upsample_nearest1d_backward
- [x] upsample_nearest2d
- [x] upsample_nearest2d_backward
- [x] upsample_nearest3d
- [x] upsample_nearest3d_backward
- [x] sigmoid_backward - [#60815]
- [x] logit_backward - [#60817]
- [x] tanh_backward - [#60816]
- [ ] slow_conv_transpose2d - [#55503, but blocked on `Optional[Tensor]` support]
- [ ] slow_conv_transpose3d
- [ ] thnn_conv2d
- [ ] thnn_conv2d_forward
- [ ] thnn_conv_depthwise2d
- [ ] thnn_conv_depthwise2d_forward
- [ ] slow_conv3d
- [ ] slow_conv3d_forward
- [ ] col2im
- [ ] col2im_backward
- [ ] column_stack
- [ ] im2col
- [ ] im2col_backward
- [ ] isposinf (#60633)
- [ ] isneginf (#60633)
- [x] special_entr https://github.com/pytorch/pytorch/pull/56082
- [ ] special_gammaln
- [ ] special_erf
- [ ] special_erfc
- [ ] special_erfinv
- [ ] fft_fft
- [ ] fft_ifft
- [ ] fft_rfft
- [ ] fft_irfft
- [ ] fft_hfft
- [ ] fft_ihfft
- [ ] fft_fft2
- [ ] fft_ifft2
- [ ] fft_rfft2
- [ ] fft_irfft2
- [ ] fft_fftn
- [ ] fft_ifftn
- [ ] fft_rfftn
- [ ] fft_irfftn
- [ ] linalg_cholesky
- [ ] linalg_slogdet
- [ ] linalg_eigh
- [ ] linalg_householder_product
- [ ] linalg_inv
- [ ] inner
- [ ] outer
- [ ] ger
- [ ] linalg_norm
- [ ] linalg_norm.ord_str
- [ ] linalg_vector_norm
- [ ] linalg_svd
- [ ] linalg_cond
- [ ] linalg_cond.p_str
- [ ] linalg_pinv
- [ ] linalg_pinv.rcond_tensor
- [ ] linalg_solve
- [ ] linalg_tensorinv
- [ ] linalg_tensorsolve
- [ ] linalg_qr
- [ ] linalg_matrix_power
- [ ] linalg_matrix_rank

Things that require adding an out kernel and then can be made structured:

- [ ] _embedding_bag (high priority)

Issues marked (XLA) are supported by XLA and thus are higher priority to port to structured.

Functions with 1d/2d/3d in their name tend to be easier and more beginner friendly

Some of these functions may not be immediately portable. Common grounds for disqualification:
- It's a reduction (blocked on #53218)
- It's multi output (need to implement this)
- It's still a TH kernel (need to do the TH to ATen port first)
- It's an alias/composite kernel (requires design from #50953)
- It behaves differently in CPU and CUDA; you'll have to fix this bug first! (Most common situation is the memory format behavior is divergent)
- It has optional tensors

### PR DESCRIPTION
**Pull request by qingyunqu**
Fixes #55070
Migrate acos to structured kernel

### CODE REVIEW

**Comment by ezyang** [comment]
Thanks @qingyunqu. This will be subsumed by https://github.com/pytorch/pytorch/pull/56082 but your contribution is still much appreciated!

**Comment by qingyunqu** [comment]
> Thanks @qingyunqu. This will be subsumed by #56082 but your contribution is still much appreciated!

ok, Thanks for review, I will close this PR.

**Comment by byronyi** [comment]
@ezyang @qingyunqu in our team are trying to move more ops into structured ones. 

Mind to suggest the next one for us to work on?

**Comment by ezyang** [comment]
@qingyunqu @bryonyi Qingyun, if you want to do the rest of the unary operations, I'm happy to let you have them; we just had an unfortunate conflict on this particular one. My preference though, is that you port all of a given class at once. https://github.com/pytorch/pytorch/pull/56151 has some test failures, do you want to fix those failures?

Besides that, there is a list of functions at https://github.com/pytorch/pytorch/issues/55070 Something that would be very helpful is for MLIR to come up with a list of priority functions, so we know which ones we should do first as opposed to later.

**Comment by ezyang** [comment]
Actually I take it back, the test failure in  #56151 is complicated; probably better to take a different class of ops.

**Comment by byronyi** [comment]
We do have a list:

```
convolution_overrideable
convolution_backward_overrideable
max_pool2d_with_indices
max_pool2d_with_indices_backward
native_batch_norm
native_batch_norm_backward
addmm
mm
_log_softmax
_log_softmax_backward_data
nll_loss_forward
nll_loss_backward
sum.dim_IntList
threshold_backward
neg
_cat
round.out
_local_scalar_dense
arange.start_out
clamp
took
index.Tensor
```
